### DISCUSSION TITLE
What is the best method to add additional Debian packages to the ISO

### DISCUSSION THREAD

**Discussion by kevinrz6** [dd010101/vyos-jenkins#49] [General]
If I needed to add a few additional Debian packages to the ISO,  like  ipset  and  iprange what would be best method.

**Comment by dd010101** [comment]
There wasn't a good method (you would need to rewrite the build-iso.sh), but now there is - see the readme for `CUSTOM_PACKAGES` env variable. You need latest version of the scripts to use this.

This applies only if you want regular package via apt install. If you want some advanced package, perhaps some kernel module that uses DKMS - this likely won't work. In such case you would need to built the kernel module manually for the specific VyOS kernel - this can't be done easily via some env variable but it can be done via the live-build hook script, specifically the `data/live-build-config/hooks/live/YOUR.chroot` bash script would allow you to do more extensive modifications inside the chroot of the ISO that is being assembled - there you can do anything you like, install via apt, build whatever, add/change/remove files, ...

**Comment by kevinrz6** [comment]
I did but I don't quite understand. How do I change the default value of  CUSTOM_PACKAGES env variable?
Do I need to modify  build-iso.sh ?  Like so ..
--custom-package vyos-1x-smoketest app1 app2 app3

**Reply by dd010101** [reply]
Call this before running build-iso.sh:

`export CUSTOM_PACKAGES="vyos-1x-smoketest emacs"`

**Comment by kevinrz6** [comment]
I've added the variables to the environment variables but the packages are not included in the ISO.

root@debian12:~# export CUSTOM_PACKAGES="vyos-1x-smoketest iprange lnav xinetd lm-sensors fancontrol i2c-tools"
root@debian12:~# env

CUSTOM_PACKAGES=vyos-1x-smoketest iprange lnav xinetd lm-sensors fancontrol i2c-tools

I've manages to successfully build, install and test standard  ISO's with no problems in my previous attempts following the instructions.
Now attempting to do a new built with the custom packages ...
Do I need to add the environment variable before any of the scripts are executed or can I just add the environment variable and just execute    build-iso.sh ?
Simply adding the variable and then executing  build-iso.sh  does not add the additional packages. Although a standard ISO is created.

Thanks
 K.

**Reply by dd010101** [reply]
Your sequence is correct. You just need to `export CUSTOM_PACKAGES=...` and then `build-iso.sh`.

Classical bash trap got me. I passed the parameter as argument without quotes and thus only first package was used and the rest was interpreted as more arguments instead of one argument with spaces. This is tricky since we passing the build command into subshell so double escaping is required for some parts.

Please download latest version of the scripts and try again, this time the `build-iso.sh` will tell you what parameters are being used so any argument/escaping error will be easy to see. Press the `y` key when running `build-iso.sh` to see the parameters.

**Comment by kevinrz6** [comment]
Downloaded the new updated build-iso.sh  and started a new build.
This time I have an error.

make: *** [Makefile:32: iso] Error 1
ISO build failed

I think there may still be an issue with quotes. Below is the initial output of the  built-iso  and I think a double quote is missing at the end of  vyos-1x-smoketest  .

Using arguments: --build-by '"kevinrz6"' --version '"1.3.8-release-20240809"' --custom-package '"vyos-1x-smoketest'


Thank you
K.

**Reply by dd010101** [reply]
I tested this before and I tested this again and it works as expected. Both branches.

```
Using arguments: --build-by 'nobody' --version '1.3.8-release-20240809' --custom-package 'vyos-1x-smoketest emacs'
```

This is expected output.

Please delete the /opt/vyos-jenkins and rerun the script to download fresh copy of everything. If you did update only build-iso.sh that's not enough.

You can make yourself update.sh, where:
```
#!/usr/bin/env bash
rm -rf /opt/vyos-jenkins
wget https://github.com/dd010101/vyos-jenkins/archive/refs/heads/master.tar.gz -O /tmp/vyos-jenkins.tar.gz
tar -xf /tmp/vyos-jenkins.tar.gz -C /tmp
mv /tmp/vyos-jenkins-master /opt/vyos-jenkins
```

The /opt/vyos-jenkins doesn't contain anything else than the scripts so you can delete it and download again without losing anything. Well only the ISO is stored there but you can create new one or simply move it somewhere else. Packages, state, progress and everything else is stored elsewhere.

**Comment by kevinrz6** [comment]
Yep,
That did it. Confirming that the custom packages were properly added to the ISO.

Thank you.

**Comment by dd010101** [comment]
👍
